### PR TITLE
lixVersions.lix_2_91: init

### PR DESCRIPTION
**Pull request by lf-**
This adds Lix 2.91.0 to nixpkgs and sets it as the default Lix release.

## Description of changes

Blog post:
https://lix.systems/blog/2024-08-12-lix-2.91-release/

Release notes:
https://docs.lix.systems/manual/lix/stable/release-notes/rl-2.91.html

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/package-management/lix/common.nix
Comment (getchoo) [inline pkgs/tools/package-management/lix/common.nix:92]:
```suggestion
  isLegacyParser = lib.versionOlder version "2.91";
```

File: pkgs/tools/package-management/lix/default.nix
Comment (getchoo) [inline pkgs/tools/package-management/lix/default.nix:37]:
```suggestion
  needsBoehmgcPatches = version: lib.versionOlder version "2.91";
```

**Review by getchoo — CHANGES_REQUESTED** [review]

File: pkgs/tools/package-management/lix/common.nix
Comment (RaitoBezarius) [inline pkgs/tools/package-management/lix/common.nix:158]: nit: i think it should go close to line 131 and maybe be `if else` ?

File: pkgs/tools/package-management/lix/common.nix
Comment (RaitoBezarius) [inline pkgs/tools/package-management/lix/common.nix:197]: hmm, i think we explicitly want to separate the debug stuff in a different output, i don't think debugoptimized does that?

Comment (lf-) [reply pkgs/tools/package-management/lix/common.nix:197]: separateDebugInfo does that. I'm not sure if our Mac builds are relatedly haunted.

File: pkgs/tools/package-management/lix/common.nix
Comment (Atemu) [reply pkgs/tools/package-management/lix/common.nix:158]: Context is easy to miss in the diff but one is buildInputs and the other nativeBuildInputs.

**Review by Atemu — APPROVED** [review]
A few nits fixed at https://github.com/lf-/nixpkgs/pull/2, please pull.

Otherwise this LGTM.

Interestingly, output path size went from 20M to 16M.

File: pkgs/tools/package-management/lix/common.nix
Comment (lf-) [reply pkgs/tools/package-management/lix/common.nix:147]: yup exactly.

**Comment by Atemu** [comment]
I always forget to mention the need for fast-forward when doing these, please rebase away the merge commit.

**Comment by lf-** [comment]
i think that's most likely because this change also fixes the LTO that was accidentally removed in a tree-wide change to, ironically, enable LTO that used to be turned off on macOS but instead mistakenly just removed the line turning it on for lix at all.

**Comment by lf-** [comment]
> I always forget to mention the need for fast-forward when doing these, please rebase away the merge commit.

yeah will do, i noticed it too late when hitting the merge button.

**Comment by lf-** [comment]
squashed and added a co-authored-by. hope that's ok for you :)

**Review by getchoo — APPROVED** [review]

**Review by RaitoBezarius — DISMISSED** [review]
It doesn't cross compile anymore to aarch64 due to missing doxygen and probably more: "kernel not defined or could not be autodetected".

**Comment by getchoo** [comment]
> It doesn't cross compile anymore to aarch64 due to missing doxygen and probably more: "kernel not defined or could not be autodetected".

This is https://git.lix.systems/lix-project/lix/issues/469, which seems to be fixed in https://gerrit.lix.systems/c/lix/+/1762. We should probably wait until that's merged upstream and pull down the patch here

**Comment by lf-** [comment]
that's strongly likely to be a nixpkgs bug in the cross file rather than a lix bug. i don't know how i fucked up pre release testing because i did test cross configurations using releaseTests in the lix flake.

**Comment by lf-** [comment]
... OHHHH i regressed cross compilation while fixing musl at the last minute after discovering they didn't implement a close_range wrapper while doing that exact testing. ok that's my own fault combined with some somewhat bad meson docs.

do y'all want me to roll a 2.91.1 release about this one line build change that fixes cross? I'm inclined to think it is appropriate because people using the overlay would need this fix to build cross as well. it'll take an hour or two but is probably the right thing to do.

**Comment by RaitoBezarius** [comment]
I think there's no hurry, take the time you need for 2.91 :-)

**Comment by lf-** [comment]
fair, i guess we probably should, in the interim, do the more expedient thing of pulling the patch because it is literally just a one line fix to the meson file.

File: pkgs/tools/package-management/lix/default.nix
Comment (getchoo) [inline pkgs/tools/package-management/lix/default.nix:67]: This is the merged patch from the gerrit ticket above
```suggestion
      docCargoHash = "sha256-KOn1fXF7k7c/0e5ZCNZwt3YZmjL1oi5A2mhwxQWKaUo=";
      
      patches = [
        # TODO: Remove next release
        # Fixes cross compilation targets
        # https://gerrit.lix.systems/c/lix/+/1762
        (fetchpatch {
          url = "https://git.lix.systems/lix-project/lix/commit/41a0b08e646b54b27ae49fb4c23e0c45c60369ca.patch";
          hash = "sha256-TZauU4RIsn07xv9vZ33amrDvCLMbrtcHs1ozOTLgu98=";
        })
      ];
```

File: pkgs/tools/package-management/lix/common.nix
Comment (getchoo) [inline pkgs/tools/package-management/lix/common.nix:125]: This should always be in `nativeBuildInputs` as it's always checked for by Meson

File: pkgs/tools/package-management/lix/common.nix
Comment (getchoo) [inline pkgs/tools/package-management/lix/common.nix:230]:
```suggestion
  postFixup = lib.optionalString enableDocumentation ''
```

$devdoc and $doc don't exist unless documentation is enabled, so this fails when it's not

**Review by getchoo — CHANGES_REQUESTED** [review]
`nix-build -A pkgsCross.aarch64-multiplatform.lix` works again after applying this diff

**Comment by kuruczgy** [comment]
(@lf-  told me to report this here)

Cross compiling lix in nixpkgs unstable is failing for me:
`mkdir: cannot create directory '/nix-support': Permission denied`

The `mkdir -p $doc/nix-support` command in the `postInstall` hook does not seem to take into account that the `doc` output is not being built on cross.

**Comment by getchoo** [comment]
That has already been fixed in this PR, as postInstall is now only used when documentation is enabled

**Comment by kuruczgy** [comment]
Ah and now I see your comment above about aarch64 cross, which is exactly what I am looking for. I will then try to use your suggested changes, thanks!

**Comment by kuruczgy** [comment]
Can confirm, with your suggestions applied (plus importing `fetchpatch` in `common.nix`) it cross-compiles successfully :+1:

File: pkgs/tools/package-management/lix/common.nix
Comment (lf-) [reply pkgs/tools/package-management/lix/common.nix:125]: nope, that's because of the internal-api-docs feature, which we should be able to disable.

File: pkgs/tools/package-management/lix/common.nix
Comment (lf-) [reply pkgs/tools/package-management/lix/common.nix:230]: oh yeah that's busted. i think i fixed it, testing it right now.

File: pkgs/tools/package-management/lix/common.nix
Comment (lf-) [reply pkgs/tools/package-management/lix/common.nix:125]: confirmed fixed.

**Comment by lf-** [comment]
@ofborg build lix

**Comment by lf-** [comment]
Fixed the musl build by adding another patch by @yu-re-ka who found I screwed up the conditional compilation that was supposed to make it work >_>

I really did not get my last-minute conditional compilation additions correct this release. Should be ok now though.

**Comment by Atemu** [comment]
@ofBorg build lix pkgsMusl.lix pkgsStatic.lix pkgsCross.aarch64-multiplatform.lix

**Review by getchoo — APPROVED** [review]
Result of `nixpkgs-review pr 334269` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>lixStatic</li>
    <li>lixStatic.dev</li>
  </ul>
</details>
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>lix (lixVersions.latest ,lixVersions.lix_2_91 ,lixVersions.stable)</li>
    <li>lix.debug (lixVersions.latest.debug ,lixVersions.lix_2_91.debug ,lixVersions.stable.debug)</li>
    <li>lix.dev (lixVersions.latest.dev ,lixVersions.lix_2_91.dev ,lixVersions.stable.dev)</li>
    <li>lix.devdoc (lixVersions.latest.devdoc ,lixVersions.lix_2_91.devdoc ,lixVersions.stable.devdoc)</li>
    <li>lix.doc (lixVersions.latest.doc ,lixVersions.lix_2_91.doc ,lixVersions.stable.doc)</li>
    <li>lix.man (lixVersions.latest.man ,lixVersions.lix_2_91.man ,lixVersions.stable.man)</li>
    <li>lixVersions.lix_2_90</li>
    <li>lixVersions.lix_2_90.debug</li>
    <li>lixVersions.lix_2_90.dev</li>
    <li>lixVersions.lix_2_90.devdoc</li>
    <li>lixVersions.lix_2_90.doc</li>
    <li>lixVersions.lix_2_90.man</li>
  </ul>
</details>

`pkgsCross.aarch64-multiplatform.lixVersions.lix_2_9{0,1}` also built

**Comment by lf-** [comment]
Going to merge this since I think it is correct now.

**Comment by lf-** [comment]
Bonus with your "fixing the darwin sandbox" content: I fixed the static build, because that was apparently easy for some reason. Sorry for all the cycles.

File: pkgs/tools/package-management/lix/common.nix
Comment (lf-) [reply pkgs/tools/package-management/lix/common.nix:192]: Nope, disabling fortify does not seem to stop LTO from exploding spectacularly in a conflagration of linker errors. I've still turned it off alongside a corrected LTO check as you suggest and pushed the result though.

**Pull request merged**
